strawberryfields/engine.py

```diff
@@ -374,19 +381,24 @@ def _run_program(self, prog, **kwargs):
             except NotApplicableError:
                 # command is not applicable to the current backend type
                 raise NotApplicableError(
-                    "The operation {} cannot be used with {}.".format(cmd.op, self.backend)
+                    f"The operation {cmd.op} cannot be used with {self.backend}."
                 ) from None
 
             except NotImplementedError:
                 # command not directly supported by backend API
                 raise NotImplementedError(
-                    "The operation {} has not been implemented in {} for the arguments {}.".format(
-                        cmd.op, self.backend, kwargs
-                    )
+                    f"The operation {cmd.op} has not been implemented in {self.backend} for the "
+                    f"arguments {kwargs}."
                 ) from None
 
         # combine the samples sorted by mode, and cast correctly to array or tensor
         samples, samples_dict = self._combine_and_sort_samples(samples_dict)
+
+        if isinstance(prog, TDMProgram) and samples_dict:
+            samples_dict = reshape_samples(samples_dict, prog.measured_modes, prog.N, prog.timebins)
+            # transpose the samples so that they have shape `(shots, spatial modes, timebins)`
```

Comment:
Why not simply have the simulator return results in the desired form? There was certainly a bit of effort on the hardware side to get them in `(shots, sptatial modes, timebins)` format.

Reply:
This only affects simulators, which return the results in the wrong order and thus need to be reshaped and transposed. So, if I recall correctly, this is needed for the results to have the same shape as the hardware.

Reply:
I understand. I just find it a little odd that it's been left to SF to get results in the right shape, rather than demanding simulators send results to SF in the desired shape.

Reply:
Oh, I see. This is mostly due to the way that TDM programs are simulated. They need to be unrolled before being run, so what is _actually_ being simulated isn't specifically a TDM circuit. The backend itself doesn't know if it's working with a TDM program or not, and thus returns the measurements in the wrong order/on the wrong modes.

I think the `_run_program` method is the deepest we can put this transformation.

Reply:
Understood.

Very cool that the backend/simulator is agnostic to the program type.